### PR TITLE
[4.0] Fix Memory Management issues in EpilogueARCAnalysis

### PR DESCRIPTION
**Pull request by gottesmm**
This sequence of relatively small commits dramatically improves the memory management in EpilogueARCAnalysis and on certain projects reduces compile time spent in the swift optimizer 36%.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

File: include/swift/SILOptimizer/Analysis/EpilogueARCAnalysis.h [hunk 1]
```diff
@@ -247,14 +247,17 @@ class EpilogueARCFunctionInfo {
...
 
-    EpilogueARCContext CM(Kind, Arg, F, PO->get(F), AA, RC->get(F));
     // Initialize and run the data flow. Clear the epilogue arc instructions if the
     // data flow is aborted in middle.
-   if (!CM.run()) { 
-     CM.resetEpilogueARCInsts();
-     return CM.getEpilogueARCInsts();
+    if (!Context.run(Kind, Arg)) {
+      Context.reset();
+      return llvm::SmallSetVector<SILInstruction *, 1>();
```
Comment (swiftix) [inline include/swift/SILOptimizer/Analysis/EpilogueARCAnalysis.h]: `return ARCInstructions();` ?

Comment (gottesmm) [inline include/swift/SILOptimizer/Analysis/EpilogueARCAnalysis.h]: I am fine making that change. But I would like to do it on master, not here. How does that sound?

Comment (swiftix) [inline include/swift/SILOptimizer/Analysis/EpilogueARCAnalysis.h]: Sounds good.

**Comment by swiftix** [comment]
LGTM! Nice work!

**Review by swiftix — APPROVED** [review]

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@shahmishal I think there may be a problem with benchmarking in swift-4.0-branch. I am only seeing one build occuring and a benchmark run from that. I am not seeing a second such build/benchmark run. The build looks for the second one, doesn't find it and it then fails.

**Comment by shahmishal** [comment]
@gottesmm @swift-ci does not support benchmark on swift-4.0-branch.

**Comment by gottesmm** [comment]
Ah. Thanks Mishal!

**Comment by gottesmm** [comment]
@swiftix I just compiled a version before/after the change of the benchmarks. They are bit-by-bit identical. So we are ok on that front.

**Comment by swiftix** [comment]
@gottesmm Thanks for going the extra mile and checking this!

**Comment by gottesmm** [comment]
Explanation: This PR improves the memory management in EpilogueARCAnalysis. This is done by drastically reducing the amount of memory allocated, using simpler data structures to manage the memory itself, and reusing already calculated information from analyses instead of recomputing. This results in certain projects improving in performance by ~30%.
Scope: This recovers more of the performance an important regression that was detected (the other commit that helped here was: 0bb34ac62568fc42fbf433200679a1b696d6756f, which is already on swift-4.0-branch).
Radar (and possibly SR Issue): rdar://33841629
Risk: There is risk to changing how we perform this memory management. On the other hand, it has been on master for a little bit and no issues have come up.
Testing: I verified that before/after the text segment of the stdlib/benchmarks are identical. So there is no change in output. I also verified by hand that we see the compile time improvement on the relevant pieces of code.

**Comment by gottesmm** [comment]
Approved by Ted via email.

**Pull request merged**